token-erc-721/chaincode-javascript/lib/tokenERC721.js

```diff
@@ -0,0 +1,484 @@
+
+    /**
+     * TokenByIndex enumerates valid non-fungible tokens.
+     *
+     * @param {Context} ctx the transaction context
+     * @param {Number} index A counter less than `totalSupply()`
+     * @returns {Number} Returns the tokenId for the index'th non-fungible token, (sort order not specified)
+     */
+    async TokenByIndex(ctx, index) {
+        const tokensAsBytes = await ctx.stub.getState(tokenArrayKey);
+        const tokens = JSON.parse(tokensAsBytes.toString());
+
+        const indexInt = parseInt(index);
+        if (indexInt >= tokens.length) {
+            throw new Error('The index is invalid. Index must be less than totalSupply()');
+        }
+
+        return tokens[indexInt];
+    }
+
+    /**
+     * Enumerate non-fungible tokens assigned to an owner.
+     *
+     * @param {Context} ctx the transaction context
+     * @param {String} owner token owner
+     * @param {Number} index A counter less than `balanceOf(owner)`
+     * @returns {Number} Returns the tokenId for the index'th non-fungible token assigned to the owner, (sort order not specified)
+     */
+    async TokenOfOwnerByIndex(ctx, owner, index) {
+        const balance = await this._readBalance(ctx, owner);
+        const indexInt = parseInt(index);
+        if (indexInt >= balance.tokens.length) {
+            throw new Error('The index is invalid. Index must be less than balanceOf(owner)');
+        }
+
+        return balance.tokens[indexInt];
+    }
+
+
+    // ============== Extended Functions for this sample ===============
+
+    /**
+     * Set optional information for a token.
+     *
+     * @param {Context} ctx the transaction context
+     * @param {String} name The name of the token
+     * @param {String} symbol The symbol of the token
+     */
+    async SetOption(ctx, name, symbol) {
+        await ctx.stub.putState(nameKey, Buffer.from(name));
+        await ctx.stub.putState(symbolKey, Buffer.from(symbol));
+        return true;
+    }
+
+    /**
+     * Mint a new non-fungible token
+     *
+     * @param {Context} ctx the transaction context
+     * @param {String} tokenId Unique ID of the non-fungible token to be minted
+     * @returns {Object} Return the non-fungible token object
+    */
+    async Mint(ctx, tokenId) {
+
+        // Check minter authorization - this sample assumes Org1 is the issuer with privilege to mint a new token
+        const clientMSPID = ctx.clientIdentity.getMSPID();
+        if (clientMSPID !== 'Org1MSP') {
+            throw new Error('client is not authorized to mint new tokens');
+        }
+
+        // Get ID of submitting client identity
+        const minter = ctx.clientIdentity.getID();
+
+        // Check if the token to be minted does not exist
+        const exists = await this._nftExists(ctx, tokenId);
+        if (exists) {
+            throw new Error(`The token ${tokenId} is already minted.`);
+        }
+
+        // Add a non-fungible token
+        const tokenIdInt = parseInt(tokenId);
```

Comment:
If you pass a token ID string that is not a number, you end up with `NaN` here and then it continues to create a token with `NaN` as the ID. We should check for errors incase anybody else makes that mistake.